langchain-core/src/prompts/chat.ts

```diff
@@ -307,6 +310,274 @@ export class ChatMessagePromptTemplate<
+    > = [];
+    for (const item of template) {
+      if (
+        typeof item === "string" ||
+        (typeof item === "object" && "text" in item)
+      ) {
+        let text = "";
+        if (typeof item === "string") {
+          text = item;
+        } else if (typeof item.text === "string") {
+          text = item.text ?? "";
+        } else {
+          // handle
+        }
+        prompt.push(PromptTemplate.fromTemplate(text));
+      } else if (typeof item === "object" && "image_url" in item) {
+        let imgTemplate = item.image_url ?? "";
+        let imgTemplateObject: ImagePromptTemplate<InputValues>;
+        let inputVariables: string[] = [];
+        if (typeof imgTemplate === "string") {
+          const parsedTemplate = parseFString(imgTemplate);
+          const variables = parsedTemplate.flatMap((item) =>
+            item.type === "variable" ? [item.name] : []
+          );
+
+          if (variables) {
+            if (variables.length > 1) {
+              throw new Error(
+                `Only one format variable allowed per image template.\nGot: ${variables}\nFrom: ${imgTemplate}`
+              );
+            }
+            inputVariables = [variables[0]];
+          } else {
+            inputVariables = [];
+          }
+
+          imgTemplate = { url: imgTemplate };
+          imgTemplateObject = new ImagePromptTemplate<InputValues>({
+            template: imgTemplate,
+            inputVariables
+          });
+        } else if (typeof imgTemplate === "object") {
+          if ("url" in imgTemplate) {
+            const parsedTemplate = parseFString(imgTemplate.url);
+            inputVariables = parsedTemplate.flatMap((item) =>
+              item.type === "variable" ? [item.name] : []
+            );
+          } else {
+            inputVariables = [];
+          }
+          imgTemplateObject = new ImagePromptTemplate<InputValues>({
+            template: imgTemplate,
+            inputVariables
+          });
+        } else {
+          throw new Error("Invalid image template");
+        }
+        prompt.push(imgTemplateObject);
+      }
+    }
+    // @ts-expect-error todo fix
+    return new this({ prompt, additionalOptions });
+  }
+
+  async format(input: TypedPromptInputValues<RunInput>): Promise<BaseMessage> {
+    // eslint-disable-next-line no-instanceof/no-instanceof
+    if (this.prompt instanceof BaseStringPromptTemplate) {
+      const text = await this.prompt.format(input);
+
+      if (this._chatMessageClass) {
+        // ChatMessage contains role field, others don't.
+        return new this._chatMessageClass({
+          content: text,
+          role: this.getRoleFromMessageClass(this._chatMessageClass.name)
+        });
+      } else if (this._messageClass) {
+        return new this._messageClass({
+          content: text
+        });
+      } else {
+        throw new Error("_messageClass and _chatMessageClass are undefined.");
```

Comment:
When would we ever get into this state?

Reply:
We need it to satisfy typescript. I pushed a refactor to make both these conditions a little cleaner